os/bluestore: list obj which equals to pend

tchaikov opened the pull request:
Fixes: https://tracker.ceph.com/issues/52705

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented:
cc @hjwsm1989

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -10878,41 +10895,30 @@ int BlueStore::_collection_list(
+          *pnext = ghobject_t::get_max();
+	} else {
+          ls->push_back(it->oid());
+          *pnext = ghobject_t::get_max();
```
not quite right, if the iterator is not hitting the 0xffffffff upper bound, in that case, we should not set the returning "next" pointer to max. 

Probably, should just take the 0xfffffff as an exception and compare the iterator with it and `pend`.

trociny replied:
Just thinking, what if instead of complicating this by adding this special handling we make sure `coll_range_temp_end` and `coll_range_end` are "the biggest" oids. I.e. modify `get_coll_range` to set not only the pool and the bitwise key, but also the namespace to some "MAX" value?

It might be a bit hackish to choose this "MAX" value for a string, but probably filling it with 0xff bytes (like `std::string(0xff, 8)`) would be good enough?

trociny replied:
Eh, I tried it and it does not seem like so simple. I have started to observe errors like
```
2021-09-24T08:08:29.580+0100 7f9630121640  0 log_channel(cluster) log [WRN] : 1.25 found stray pgmeta-like #1:a8000000::::head# during PG removal
```
Which I think are related.

trociny replied:
Thinking a little more about it. I suppose I had to modify `coll_range_temp_end` to set "MAX" namespace only for the case when `end_hash == 0xffffffffull`.  For the `end_hash` we have:
```
    uint64_t end_hash = reverse_hash  + (1ull << (32 - bits));
    if (end_hash > 0xffffffffull)
      end_hash = 0xffffffffull;
```
I.e. normally `end_hash` is out of the range, and `end` oid built with this hash will be always the biggests. But when we overflow (`end_hash` becomes `0xffffffffull + 1`) we set it to 0xffffffffull and then have this weird things.

At least this is my interpretation. And I tried this patch and it seems to work for me (though not tested well):
```
--- a/src/os/bluestore/BlueStore.cc
+++ b/src/os/bluestore/BlueStore.cc
@@ -311,6 +311,11 @@ static void get_coll_range(const coll_t& cid, int bits,
 
     end->hobj.set_bitwise_key_u32(end_hash);
     temp_end->hobj.set_bitwise_key_u32(end_hash);
+
+    if (end_hash == 0xffffffffull) {
+      end->hobj.nspace = std::string(0xff, 8);
+      temp_end->hobj.nspace = std::string(0xff, 8);
+    }
   } else {
     start->shard_id = shard_id_t::NO_SHARD;
     start->hobj.pool = -1ull;
@@ -322,6 +327,9 @@ static void get_coll_range(const coll_t& cid, int bits,
     // no separate temp section
     *temp_start = *end;
     *temp_end = *end;
+
+    end->hobj.nspace = std::string(0xff, 8);
+    temp_end->hobj.nspace = std::string(0xff, 8);
   }
 
   start->generation = 0;
```
And I am not telling this is the right solution.

hjwsm1989 replied:
i think this will work, one little nit, why std::string(0xff, 8), should  std::string(0xff, 4) instead？

trociny replied:
`std::string(0xff, 8)` was just the first thing that came to my mind. I suppose even just `std::string(0xff, 1)`  would be enough.

trociny reviewed (commented):
It would be nice to add a test for this 0xffffffff hash age case into test/objectstore/store_test.cc.

trociny commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -306,18 +307,22 @@ static void get_coll_range(const coll_t& cid, int bits,
     *end = *start;
     start->hobj.set_bitwise_key_u32(0);
-    end->hobj.set_bitwise_key_u32(0xffffffff);
+    end->hobj.set_bitwise_key_u32(MAX_HASH);
```
I have not checked where this "no shard" case may be actually used, so just wonder if you are sure we don't need the same trick here?

tchaikov replied:
@trociny yup. i don't think there is any guarantee this won't happen when it comes to meta PG. so i am applying the same trick here. also, the tests are added.

hjwsm1989 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -306,19 +307,23 @@ static void get_coll_range(const coll_t& cid, int bits,
-
-    end->hobj.set_bitwise_key_u32(end_hash);
-    temp_end->hobj.set_bitwise_key_u32(end_hash);
+    if (end_hash > MAX_HASH) {
```
what about the end_hash == MAX_HASH  case?

trociny reviewed (approved):
LGTM

xiexingguo commented:
@tchaikov @trociny Does this problem exist on nautilus and prior versions?

tchaikov commented:
> @tchaikov @trociny Does this problem exist on nautilus and prior versions?

yup, will backport this fix once it is merged.

xiexingguo reviewed (approved):
Thanks for catching this!

ideepika commented:
```
2021-09-27T17:45:55.502 INFO:teuthology.orchestra.run.smithi183.stderr:Cleaning
2021-09-27T17:45:55.959 INFO:teuthology.orchestra.run.smithi183.stdout:==> rm -r bluestore.test_temp_dir
2021-09-27T17:45:55.963 INFO:teuthology.orchestra.run.smithi183.stdout:[       OK ] ObjectStore/StoreTest.ListEndTest/2 (4232 ms)
2021-09-27T17:45:55.963 INFO:teuthology.orchestra.run.smithi183.stdout:[ RUN      ] ObjectStore/StoreTest.List_0xfffffff_Hash_Test_in_meta/2
2021-09-27T17:45:59.630 INFO:teuthology.orchestra.run.smithi183.stdout:/build/ceph-17.0.0-7835-g52b2f8d5/src/test/objectstore/store_test.cc:2959: Failure
2021-09-27T17:45:59.631 INFO:teuthology.orchestra.run.smithi183.stdout:Expected equality of these values:
2021-09-27T17:45:59.631 INFO:teuthology.orchestra.run.smithi183.stdout:  objects.size()
2021-09-27T17:45:59.631 INFO:teuthology.orchestra.run.smithi183.stdout:    Which is: 0
2021-09-27T17:45:59.631 INFO:teuthology.orchestra.run.smithi183.stdout:  1
```
@tchaikov can you take a look http://qa-proxy.ceph.com/teuthology/ideepika-2021-09-27_14:52:49-rados-wip-deepika-testing-2021-09-27-1601-distro-basic-smithi/6409464/teuthology.log

trociny commented:
@tchaikov I have not checked it but I have a suspicion that the failure is when "legacy" version of collection_list is used. The test's `collection_list` function is designed to randomly chose "legacy" and "normal" version, if `disable_legacy` param is not set.

Right now I am not quite sure why the legacy would not work, may be due to how the "0xff" namespace is encoded into the key? If it is the case I am not sure we want to fix the "legacy" version, may be just disable it in the test. "legacy" is supposed to behave the same way as at is in "old" bluestore osds. The osd uses it when the pg active set contains a mix of "old" and "new" osds.

tchaikov commented:
@ideepika sure. fixed and repushed.

@trociny  you are right. i am able to reproduce the failure if the "legacy" branches is always enabled in `collection_list()`. but i just fixed the legacy case as well just in case.

changelog:

- use `\0x7f` as the nspace of the upper bound hobj if listing in the legacy mode. this should address the test failure of http://qa-proxy.ceph.com/teuthology/ideepika-2021-09-27_14:52:49-rados-wip-deepika-testing-2021-09-27-1601-distro-basic-smithi/6409464/teuthology.log

tchaikov commented:
- https://pulpito.ceph.com/kchai-2021-09-28_23:40:16-rados-wip-kefu-testing-2021-09-28-2248-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-09-29_14:47:07-rados-wip-kefu-testing-2021-09-28-2248-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/52764
- https://tracker.ceph.com/issues/52765
